### DISCUSSION TITLE
Logging in to Cosmic Verge

### DISCUSSION THREAD

**Discussion by ecton** [khonsulabs/cosmicverge#16] [Ideas]
We are attempting to create a painless but secure process around logging in and recovering accounts in Cosmic Verge. This is what I remember from the voice conversation we had today.

## Account Recovery Code

Upon logging in for the first time, a unique backup code will be generated and presented to the user as an Account Recovery Code. This code will be alpha-numeric and will be presented to the user as their method of recovering their account if they lose access to their normal authentication methods.

The idea behind this code is that it shouldn't be used on a regular basis, and because it's server-side-generated, we hope most average users will write it somewhere safe, and never use it. But, if they are no longer able to access their account, they can use this code to recover their account. Because it's server-side generated, we can be confident it doesn't exist in a known database of leaked passwords and meets specific security requirements.

## Primary Authentication Providers

Primary Authentication Providers are mechanisms of validating who the user is. These providers are the methods we're wanting to support for logging into the game:

* OAuth2/OpenID Connect providers
* Verified Email via Magic Link or Code (code being hand-entered when on a device that doesn't have access to email)
* Client Certificate (not for the general public)

We will require email verification to play the game, but until we have a privacy policy, we should avoid collecting email addresses.

## Secondary Authentication Mechanisms

Secondary Authentication Mechanisms are methods to validate that the user that the Primary Authentication Provider says is logging in is actually logging in. Without this step, if a user's Twitch credentials were hacked, a player could use Twitch to log in without verifying that they are who they say they are. The available methods should include:

* Email a time-based code (But only if Magic Link wasn't used as the Primary Authentication Provider)
* SMS a time-based code
* TOTP a la Google Authenticator
* [FIDO2/CTAP2-compatible](https://crates.io/crates/authenticator) hardware device

We are 100% confident that we want all superuser accounts to require either TOTP or a hardware key. What we are unsure of is whether we can reasonably force users to enable 2-factor or whether it will become a support nightmare. I think we all would like to ensure two-factor is enabled for as many people as possible, but the reality is that many users can be easily confused setting these tools up.

**Comment by daxpedda** [comment]
Activating 2FA requires setting up a password and SMS.
The reason being that these are recoverable, all others aren't.

You can't setup a password without 2FA, and you can't use SMS for authentification alone, both of these are only used together in a recovery scenario, one of them isn't enough to recover.
